kv/memberlist/tcp_transport.go

```diff
@@ -378,12 +378,25 @@ func (t *TCPTransport) FinalAdvertiseAddr(ip string, port int) (net.IP, int, err
 
 			advertiseAddr = net.ParseIP(ip)
 			if advertiseAddr == nil {
-				return nil, 0, fmt.Errorf("failed to parse advertise address: %q", ip)
+				return nil, 0, fmt.Errorf("failed to parse advertise address %q", ip)
 			}
-		} else {
+		case colonColonZero:
+			inet6Ip, err := netutil.GetFirstAddressOf(nil, t.logger, true)
+			if err != nil {
+				return nil, 0, fmt.Errorf("failed to get private inet6 address: %w", err)
+			}
+
+			advertiseAddr = net.ParseIP(inet6Ip)
+			if advertiseAddr == nil {
+				return nil, 0, fmt.Errorf("failed to parse inet6 advertise address %q", ip)
+			}
+		default:
 			// Use the IP that we're bound to, based on the first
 			// TCP listener, which we already ensure is there.
 			advertiseAddr = t.tcpListeners[0].Addr().(*net.TCPAddr).IP
+			if advertiseAddr == nil {
+				return nil, 0, fmt.Errorf("failed to parse advertise address from tcp listener %q", ip)
+			}
```

Comment:
[nit] There's no parsing here. The IP is guaranteed to be returned from the TCP listener. I think we can remove this new error check.